### PR TITLE
Track local checkpoint on primary immediately

### LINKED ISSUE #25415: After recovering, a primary should update its knowledge of its own local checkpoint
A primary shard should update its knowledge of its own local checkpoint so that the global checkpoint does not otherwise know the local checkpoint of the shard until an indexing operation.

Relates #10708, relates #25355

### PR DESCRIPTION
**Pull request by jasontedor**
Today after a primary shard is recovered or promoted from a replica, it will not be tracking its own local checkpoint until the first indexing operation on the shard. This leads to situations where a shard is relocated and the knowledge it sends in the primary context about its own local checkpoint violates the global checkpoint. This commit rectifies this situation by tracking the local checkpoint on such shards immediately on recovery/promotion/relocation.

Closes #25415, relates #10708, relates #25355

### CODE REVIEW

**Comment by jasontedor** [comment]
@ywelsch I'm opening this for discussion.

**Comment by jasontedor** [comment]
@ywelsch I think this is ready for you after integrating #25431.

File: core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointTracker.java [hunk 1]
```diff
@@ -360,6 +365,7 @@ synchronized void updateAllocationIdsFromPrimaryContext(final PrimaryContext pri
                 .allMatch(e -> e.value == SequenceNumbersService.UNASSIGNED_SEQ_NO) : inSyncLocalCheckpoints;
         assert StreamSupport
                 .stream(trackingLocalCheckpoints.spliterator(), false)
+                .filter(e -> !e.key.equals(allocationId)) // during primary relocation a shard can already know its local checkpoint
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointTracker.java]: I think we should not filter out the allocation id, but keep the method as before and instead assert in the `IndexShard.updateAllocationIdsFromPrimaryContext` method that the local checkpoint that is provided as part of the primary context is equal to the local checkpoint that exists locally.

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 2]
```diff
@@ -374,8 +375,12 @@ public void updateShardState(final ShardRouting newRouting,
                 final Engine engine = getEngineOrNull();
                 // if the engine is not yet started, we are not ready yet and can just ignore this
                 if (engine != null) {
-                    engine.seqNoService().updateAllocationIdsFromMaster(
-                            applyingClusterStateVersion, activeAllocationIds, initializingAllocationIds);
+                    final SequenceNumbersService seqNoService = engine.seqNoService();
+                    seqNoService.updateAllocationIdsFromMaster(applyingClusterStateVersion, activeAllocationIds, initializingAllocationIds);
+                    if ((currentState == IndexShardState.POST_RECOVERY && state == IndexShardState.STARTED) ||
+                            recoveryState.getRecoverySource().getType().equals(RecoverySource.Type.PEER)) {
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: With the change I suggested above we can leave out the == PEER condition and instead add ` && currentRouting.isRelocationTarget == false` to `currentState == IndexShardState.POST_RECOVERY && state == IndexShardState.STARTED`

File: core/src/test/java/org/elasticsearch/index/shard/LocalCheckpointIT.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+import java.util.concurrent.ExecutionException;
+
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
+import static org.hamcrest.Matchers.equalTo;
+
+/**
+ * Tests that a primary shard tracks its own local checkpoint after starting.
+ */
+@ESIntegTestCase.ClusterScope(scope = Scope.TEST, numDataNodes = 0)
+public class LocalCheckpointIT extends ESIntegTestCase {
```
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/index/shard/LocalCheckpointIT.java]: I wonder if we need an integration test for this. A test under `IndexShardTests` would be able to check the same.

**Review by ywelsch — CHANGES_REQUESTED** [review]
I've made a suggestion around the update condition and suggested to use unit tests instead.

File: core/src/test/java/org/elasticsearch/index/shard/LocalCheckpointIT.java (hunk 3 above)
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/index/shard/LocalCheckpointIT.java]: Yeah, now that we are not doing this via indices cluster state service, this is now possible. I will update.

**Comment by jasontedor** [comment]
This will be superseded by a forthcoming PR.

**Pull request closed without merging**